Gecko EMF Exporters (CSV, XLSX, ODS) enhancements

ideas-into-software opened the pull request:
- "Model Metadata miss isID" (Issue #32): https://github.com/geckoprojects-org/org.gecko.emf.utils/issues/32

 - "Pseudo ID identification an handling" (Issue #33): https://github.com/geckoprojects-org/org.gecko.emf.utils/issues/33

 - "Inner document non containment references" (Issue #34): https://github.com/geckoprojects-org/org.gecko.emf.utils/issues/34

 - "PseudoID if non ID field named id is present" (Issue #35): https://github.com/geckoprojects-org/org.gecko.emf.utils/issues/35
 
 - Various other fixes & improvements

Commit:
Gecko EMF Exporters (CSV, XLSX, ODS)  (c.d.)

 - "Model Metadata miss isID" (Issue #32)

 - "Pseudo ID identification an handling" (Issue #33)

 - "Inner document non containment references" (Issue #34)

 - "PseudoID if non ID field named id is present" (Issue #35)

 - Various other fixes & improvements

Signed-off-by: Michael H. Siemaszko <m.siemaszko@datainmotion.com>

Commit:
Gecko EMF Exporters (CSV, XLSX, ODS)  (c.d.)

 - Temporarily disable test case until issue is clarified

Signed-off-by: Michael H. Siemaszko <m.siemaszko@datainmotion.com>

ideas-into-software commented:
@juergen-albert

PR is in draft mode until following is clarified: 

Side effect of https://github.com/geckoprojects-org/org.gecko.emf.utils/issues/35 causes a problem in models which have fields called ID but not marked as ID (e.g. `Request` in `org.gecko.emf.util.model`). This results is 2 (TWO) ID fields in output generated, in addition to "internal id" ("_id") field.
	    	
How should such cases be handled ? There CANNOT be two columns with same name (i.e. "ID" column more than once), as this is not compatible with some of the already existing formats (e.g. R language data frame). 

Beside clarifying above mentioned, please inspect output produced by this latest code and let me know if there's anything else on this topic (EMF Exporters).

![Screenshot from 2023-12-09 02-08-03](https://github.com/geckoprojects-org/org.gecko.emf.utils/assets/9023040/f8a05372-ef10-48b7-846b-4f564a1a25bc)
![Screenshot from 2023-12-09 02-07-51](https://github.com/geckoprojects-org/org.gecko.emf.utils/assets/9023040/c01a7429-3b6a-404b-881f-e8893b7eb4e9)

juergen-albert commented:
I'm confused, why there are suddenly 2 fields named ìd`!?

My intention in general and with #35 was:

1. There is always a field `_id`.
2. If there is an Attribute where `isID` is true the value of that attribute will be the value of `_id`
3. If there is no Attribute where `isID` is true the value of `_id` will be the generated pseudo ID
4. An attribute where the name is `name.toLowerCase().equals("id")` and `isID` is `false` must not be used to populate the `_id` field.

ideas-into-software commented:
Hi Juergen,

Please re-read my earlier comment, as it appears you may have overlooked it: https://github.com/geckoprojects-org/org.gecko.emf.utils/pull/36#issuecomment-1848800127

What you write, i.e.: 
> 
> 1. There is always a field `_id`.
> 2. If there is an Attribute where `isID` is true the value of that attribute will be the value of `_id`
> 3. If there is no Attribute where `isID` is true the value of `_id` will be the generated pseudo ID
> 4. An attribute where the name is `name.toLowerCase().equals("id")` and `isID` is `false` must not be used to populate the `_id` field.

.. is how this works already. 


Problem is with models which have fields called ID but not marked as ID (e.g. `Request` in `org.gecko.emf.util.model`); quoting my earlier comment (https://github.com/geckoprojects-org/org.gecko.emf.utils/pull/36#issuecomment-1848800127): 

> Side effect of #35 causes a problem in models which have fields called ID but not marked as ID (e.g. Request in org.gecko.emf.util.model). This results is 2 (TWO) ID fields in output generated, in addition to "internal id" ("_id") field.
> 
> How should such cases be handled ? There CANNOT be two columns with same name (i.e. "ID" column more than once), as this is not compatible with some of the already existing formats (e.g. R language data frame).

Please address that part so I know how to handle these cases, i.e. what I asked about originally.

Commit:
Gecko EMF Exporters (CSV, XLSX, ODS)  (c.d.)

 - "PseudoID if non ID field named id is present (Issue #35)" (c.d.)

 - "rename Metadata Sheets/fields" (Issue #37)

 - "Class documentation missing" (Issue #38)
 
 - "Help resolver to pull in the required bundles" (Issue #39)
 

Signed-off-by: Michael H. Siemaszko <m.siemaszko@datainmotion.com>

juergen-albert commented on org.gecko.emf.csv/src/org/gecko/emf/csv/configuration/EMFCSVResourceFactory.java:
```diff
@@ -14,13 +14,20 @@
  * @author Michal H. Siemaszko
  */
+@Requirement(namespace = EMFExporterConstants.EMF_EXPORTER_NAMESPACE, name = EMFCSVExporterConstants.EMF_EXPORTER_NAME)
+@Capability(namespace = EMFCSVConstants.EMF_RESOURCEFACTORY_NAMESPACE, name = EMFCSVConstants.EMF_RESOURCEFACTORY_NAME)
```
The Provide is not technically needed, as the EMFCSVResourceFactoryConfigurator already provides a unique capability.

juergen-albert commented on org.gecko.emf.csv/src/org/gecko/emf/csv/annotation/RequireEMFCSVResourceFactory.java:
```diff
@@ -0,0 +1,39 @@
+@Documented
+@Retention(CLASS)
+@Target({ TYPE, PACKAGE })
+@Requirement(namespace = EMFCSVConstants.EMF_RESOURCEFACTORY_NAMESPACE, name = EMFCSVConstants.EMF_RESOURCEFACTORY_NAME)
```
You could also require the already  existing Capability on  EMFCSVResourceFactoryConfigurator.

juergen-albert commented on org.gecko.emf.csv.tests/test.bndrun:
```diff
@@ -5,6 +5,15 @@
 
 -runbundles.junit5: ${test.runbundles}
 
+-runrepos: \
```
I can't add a comment to the line in the file, I want to so I add it here:

The Test if the Requirements work correctly would be, that the only runrequirement defined here is the `org.gecko.emf.csv`.tests. As it references the `EMFCSVResource`, everything required should be resolved by itself.

ideas-into-software replied:
@juergen-albert 

This was tested already via `/org.gecko.emf.util.jakartars.bnd.library.workspace/required.bndrun` and `/org.gecko.emf.util.jaxrs.bnd.library.workspace/required.bndrun` and worked with no problems.

However, the test which you suggested does not work. What happens is a non-deterministic behavior where additional, unexpected exporter bundles are pulled in during resolve, i.e. when resolving via `/org.gecko.emf.csv.tests/test.bndrun`, having defined only `org.gecko.emf.csv.tests` in its run requirements, in addition to all of the expected bundles (`org.gecko.emf.csv`, `org.gecko.emf.csv.tests`, `org.gecko.emf.exporter`, `org.gecko.emf.exporter.csv`, `org.gecko.emf.exporter.csv.api`), an unexpected bundle appears in resolution results – unexpected in the sense that it is not specific to this format – along with its transitive dependencies (e.g. `org.gecko.emf.exporter.ods` with `com.github.miachm.sods` in some cases, `org.gecko.emf.exporter.xlsx` with `org.apache.servicemix.bundles.poi` in other cases). Out of 8 `test.bndrun`s, this behavior appears in 3 (`/org.gecko.emf.xlsx.tests/test.bndrun`, `/org.gecko.emf.r_lang.tests/test.bndrun`, `/org.gecko.emf.csv.tests/test.bndrun`) and does not appear in the rest (`/org.gecko.emf.ods.tests/test.bndrun`, `/org.gecko.emf.exporter.xlsx.tests/test.bndrun`, `/org.gecko.emf.exporter.r_lang.tests/test.bndrun`, `/org.gecko.emf.exporter.ods.tests/test.bndrun`, `/org.gecko.emf.exporter.csv.tests/test.bndrun`).

It appears this problem is caused by lack of sufficient information for resolver to determine which specific format of EMF exporter to pull in when detecting OSGi service requirement defined via `org.osgi.service.component.annotations.Reference` – e.g. in `org.gecko.emf.csv.EMFCSVResourceFactoryConfigurator`, `org.gecko.emf.exporter.EMFExporter` is referenced, but even though its target property is correctly specified (`(component.name=EMFCSVExporter)`) this information is not passed on to manifest headers generated, i.e. `Require-Capability` manifest header only defines `osgi.service;filter:="(objectClass=org.gecko.emf.exporter.EMFExporter)";effective:=active` which is too generic. 

IMHO it would be useful to understand underlying cause of such non-deterministic behavior, similar non-deterministic behavior mentioned in other question asked on Slack several days ago, which was left without answer (i.e. imports generated only in some bundles, even though all of them use the same pattern in the sense of importing constants class).

![Screenshot from 2023-12-26 02-36-00](https://github.com/geckoprojects-org/org.gecko.emf.utils/assets/9023040/c5fd83be-f352-4aa6-ac6b-0ad6350c0b3e)

juergen-albert commented on org.gecko.emf.exporter.ods/bnd.bnd:
```diff
@@ -13,3 +13,7 @@ Bundle-Description: ODS (OpenDocument Spreadsheet) Exporter for EMF
 
 Private-Package: org.gecko.emf.exporter.ods.impl
+
+Import-Package: \
```
`Import-Package` can be removed, as the Bundle you add manually only appear to contain Constants.

juergen-albert replied:
The Private-Package can be removed as well, as the bnd default will package everything correctly.

juergen-albert commented:
If I'm not mistaken, my comments around the CSV Exporter EMFResource apply to XLS, ODS and R as well.

Commit:
Gecko EMF Exporters (CSV, XLSX, ODS)  (c.d.)

 - Additional "no options" test cases
 
 - CSV Exporter - default to ZIP export mode if no options are specified
 
 - Various fixes and improvements for Requirements / Capabilities (Issue 
#39) and other


Signed-off-by: Michael H. Siemaszko <m.siemaszko@datainmotion.com>

ideas-into-software commented:
@juergen-albert 

I would like to ask that you clarify as what you requested is logically incoherent; let’s trace this back:

**A**. In https://github.com/geckoprojects-org/org.gecko.emf.utils/issues/39, you asked me for two things – quoting: 
 - part no. 1: “_Individual Exporters should provide a capability that is in return required by there emf counterparts that provide the ResourceFactories._”
 - part no. 2: “_As the bundles with the resource Factories are usually very indirect dependencies, they should provide a Capability as well, together with a RequireCapability Annotation like RequireEMFJson._”

**B**. Then, as part of  commit 89ccb8dd3bc074bf6e381fd18cdf787b163afb55 (https://github.com/geckoprojects-org/org.gecko.emf.utils/pull/36/commits/89ccb8dd3bc074bf6e381fd18cdf787b163afb55), that is exactly what was delivered, via the following: 
 - As per part no. 1 of https://github.com/geckoprojects-org/org.gecko.emf.utils/issues/39: format-specific EMF Exporter capability (`org.osgi.annotation.bundle.Capability`) defined in each of the individual exporters, i.e. `org.gecko.emf.exporter.csv.impl.EMFCSVExporter`, `org.gecko.emf.exporter.ods.impl.EMFODSExporter`, `org.gecko.emf.exporter.r_lang.impl.EMFRLangExporter`, `org.gecko.emf.exporter.xlsx.impl.EMFXLSXExporter`;
  - As per part no. 1 of https://github.com/geckoprojects-org/org.gecko.emf.utils/issues/39: requirement (`org.osgi.annotation.bundle.Requirement`) of format-specific EMF Exporter capability defined in respective resource factories, i.e. `org.gecko.emf.csv.configuration.EMFCSVResourceFactory`, `org.gecko.emf.ods.configuration.EMFODSResourceFactory`, `org.gecko.emf.r_lang.configuration.EMFRLangResourceFactory`, `org.gecko.emf.xlsx.configuration.EMFXLSXResourceFactory`;
 - As per part no. 2 of https://github.com/geckoprojects-org/org.gecko.emf.utils/issues/39: format-specific EMF Resource Factory capability (`org.osgi.annotation.bundle.Capability`) defined in respective resource factories, i.e.: `org.gecko.emf.csv.configuration.EMFCSVResourceFactory`, `org.gecko.emf.ods.configuration.EMFODSResourceFactory`, `org.gecko.emf.r_lang.configuration.EMFRLangResourceFactory`, `org.gecko.emf.xlsx.configuration.EMFXLSXResourceFactory`;
 - As per part no. 2 of https://github.com/geckoprojects-org/org.gecko.emf.utils/issues/39: requirement (`org.osgi.annotation.bundle.Requirement`) of format-specific EMF Resource Factory capability defined via additional annotations like `RequireEMFJson`, i.e.: `org.gecko.emf.csv.annotation.RequireEMFCSVResourceFactory`, `org.gecko.emf.ods.annotation.RequireEMFODSResourceFactory`, `org.gecko.emf.r_lang.annotation.RequireEMFRLangResourceFactory`, `org.gecko.emf.xlsx.annotation.RequireEMFXLSXResourceFactory`;

**C**. Since it was not clear to me why part no. 2 of https://github.com/geckoprojects-org/org.gecko.emf.utils/issues/39 is actually needed, on 2023/12/19 I asked you a question via Slack (https://dim-c.slack.com/archives/D01TP6194TZ/p1702983714503899): “_it is not clear to me why another capability is needed (i.e. in `org.gecko.emf.csv.configuration.EMFCSVResourceFactory`, `org.gecko.emf.ods.configuration.EMFODSResourceFactory`, `org.gecko.emf.r_lang.configuration.EMFRLangResourceFactory`, `org.gecko.emf.xlsx.configuration.EMFXLSXResourceFactory`) if `org.gecko.emf.csv.EMF*ResourceFactoryConfigurator` already defines such for EMF CSV / ODS / XLSX / R LANG._”

**D**. I never received a reply to that question (nor the other question I asked in same Slack message), instead, in code review, you mentioned: 
 - “_The Provide is not technically needed, as the EMFCSVResourceFactoryConfigurator already provides a unique capability._” (https://github.com/geckoprojects-org/org.gecko.emf.utils/pull/36#pullrequestreview-1790878243)
 - “_You could also require the already existing Capability on EMFCSVResourceFactoryConfigurator._” (https://github.com/geckoprojects-org/org.gecko.emf.utils/pull/36#pullrequestreview-1790879970)

**E**. … which is in conflict with what you asked in part no. 2 of https://github.com/geckoprojects-org/org.gecko.emf.utils/issues/39 AND would result in having an exact duplicate of already existing annotations, i.e.: `org.gecko.emf.csv.annotation.RequireEMFCSV`, `org.gecko.emf.ods.annotation.RequireEMFODS`, `org.gecko.emf.r_lang.annotation.RequireEMFRLang`, `org.gecko.emf.xlsx.annotation.RequireEMFXLSX`;

juergen-albert commented:
Okay, lets boil it down with the example of the CSV Exporter. I want for each Exporter the following:

Annotation `RequireEMFCSV` - This is the only thing I want to add to my code, when I need a CSV Export. This should lead the resolver to the `org.gecko.emf.csv` bundle. It already filters for the `ProvideEMFResourceConfigurator` on the `EMFCSVResourceFactoryConfigurator`, so this is already there.

The `EMFCSVResourceFactoryConfigurator` has a reference to the `EMFExporter` service (let's keep in mind, that the resolver will not know about the target filter).  Thus the  `org.gecko.emf.csv` bundle will need another requirement, that the must be stratified by some Capability the `EMFCSVExporter`s Bundle provides. 

This is all, I wanted.
